File: javascript/src/algorithms/reverse_string.js
import { benchmark, print_recap } from "../tools.js";

// Premise :
// You receive a string
// You return this string, reversed


class reverse_string_t {
    constructor() {
        // this.iterations = 500000;
        this.iterations = 1;
    }

    method_0(str) {
        return str.split("").reverse().join("");
    }

    method_1(str) {
        let reversed = "";
        const l = str.length;
        for (let i = l-1; i >= 0; --i) {
            reversed = `${reversed}${str[i]}`;
        }
        return reversed;
    }

    method_2(str) {
        let reversed = "";
        for (const character of str) {
            reversed = character + reversed;
        }
        return reversed;
    }

    // Fastest
    method_3(str) {
        return str.split("").reduce((reversed, character) => character + reversed, "");
    }

    method_4(str) {
        return str.split("").reduce((rev, char) => `${char}${rev}`, "");
    }

    run_batch() {
        const input = "Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim. Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim. Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim. Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim. Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim. Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim. Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim. Voluptate elit duis reprehenderit in et in nisi elit dolore non mollit anim."

        let res = {};
        res[0] = benchmark("0", this.method_0, input, this.iterations);
        res[1] = benchmark("1", this.method_1, input, this.iterations);
        res[2] = benchmark("2", this.method_2, input, this.iterations);
        res[3] = benchmark("3", this.method_3, input, this.iterations);
        res[4] = benchmark("4", this.method_4, input, this.iterations);

        print_recap(res);
    }
}

let reverse_string = new reverse_string_t();

export { reverse_string };

File: python/src/algorithms/reverse_string.py
from ..benchmark import execute_test, print_report
import functools

class reverse_string_t(object):
    def __init__(self):
        # self.iterations = 900000
        self.iterations = 1
    
    def method_0(self, str):
        return str[::-1]

    def method_1(self, str):
        return ''.join(reversed(str))

    def method_2(self, str):
        return functools.reduce(lambda res, char: f"{char}{res}", str, "")

    def run_batch(self):
        inputs = "Baked beans pie with hot sauce cider in a cupcake paper cup."

        res = []
        res.append(execute_test("0", self.method_0, inputs, self.iterations))
        res.append(execute_test("1", self.method_1, inputs, self.iterations))
        res.append(execute_test("2", self.method_2, inputs, self.iterations))

        print_report(res)

reverse_string = reverse_string_t()
